File: ltms/go/marx.go
// Marx Brothers puzzle.
// Converted from marx.lisp.
package ltms

import "fmt"

var marxAttributes = []string{
	"PLAYS-PIANO", "PLAYS-HARP", "SMOOTH-TALKER",
	"LIKES-GAMBLING", "LIKES-ANIMALS",
}

var marxObjects = []string{"GROUCHO", "HARPO", "CHICO"}

// MarxBrothers solves the Marx Brothers attribution problem.
func MarxBrothers() {
	SolveAttributionProblem(marxAttributes, marxObjects)
}

// MakeAttributeChoiceSets creates choice sets for DD-Search.
func MakeAttributeChoiceSets(attributes, objects []string) [][]interface{} {
	var result [][]interface{}
	for _, attr := range attributes {
		var choices []interface{}
		for _, obj := range objects {
			choices = append(choices, []interface{}{attr, obj})
		}
		result = append(result, choices)
	}
	return result
}

// SolveAttributionProblem solves an attribution problem using DD-Search.
func SolveAttributionProblem(attributes, objects []string) {
	InLTRE(CreateLTRE("Attribution Problem Scratchpad", false))
	LoadMarxData(CurrentLTRE)
	DDSearch(
		MakeAttributeChoiceSets(attributes, objects),
		func() { ShowAttributeSolution(attributes) },
		CurrentLTRE,
	)
}

// ShowAttributeSolution displays the solution.
func ShowAttributeSolution(attributes []string) {
	fmt.Println("\nSolution:")
	for _, attr := range attributes {
		matches := Fetch([]interface{}{attr, "?object"}, CurrentLTRE)
		for _, match := range matches {
			if IsTrue(match, CurrentLTRE) {
				fmt.Printf("\n  %v", match)
			}
		}
	}
}

File: ltms/java/Marx.java
package ltms;

import java.util.*;

public class Marx {
    public static final String[] ATTRIBUTES = {
        "PLAYS-PIANO", "PLAYS-HARP", "SMOOTH-TALKER",
        "LIKES-GAMBLING", "LIKES-ANIMALS"
    };

    public static final String[] OBJECTS = {"GROUCHO", "HARPO", "CHICO"};

    public static void marxBrothers() {
        solveAttributionProblem(ATTRIBUTES, OBJECTS);
    }

    public static List<List<Object>> makeAttributeChoiceSets(String[] attributes, String[] objects) {
        List<List<Object>> result = new ArrayList<>();
        for (String attr : attributes) {
            List<Object> choices = new ArrayList<>();
            for (String obj : objects) {
                choices.add(Arrays.asList(attr, obj));
            }
            result.add(choices);
        }
        return result;
    }

    public static void solveAttributionProblem(String[] attributes, String[] objects) {
        LTRE.inLTRE(LTRE.createLTRE("Attribution Problem Scratchpad", false));
        MarxData.loadMarxData(LTRE.currentLTRE);
        DDS.ddSearch(
            makeAttributeChoiceSets(attributes, objects),
            () -> showAttributeSolution(attributes),
            LTRE.currentLTRE
        );
    }

    public static void showAttributeSolution(String[] attributes) {
        System.out.println("\nSolution:");
        for (String attr : attributes) {
            List<Object> matches = LTRE.fetch(Arrays.asList(attr, "?object"), LTRE.currentLTRE);
            for (Object match : matches) {
                if (LTRE.isTrue(match, LTRE.currentLTRE)) {
                    System.out.printf("\n  %s", match);
                }
            }
        }
    }
}
